lib/charms/operator_libs_linux/v0/dnf.py

```diff
@@ -0,0 +1,207 @@
+# Copyright 2023 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Abstractions for system's DNF package information and repositories."""
+
+import os
+import re
+import shutil
+import subprocess
+from dataclasses import dataclass
+from enum import Enum
+from typing import Optional, Union
+
+
+class Error(Exception):
+    """Raise when dnf encounters an execution error."""
+
+
+class _PackageState(Enum):
+    INSTALLED = "installed"
+    AVAILABLE = "available"
+    ABSENT = "absent"
+
+
+@dataclass(frozen=True)
+class PackageInfo:
+    """Dataclass representing DNF package information."""
+
+    name: str
+    arch: str = None
+    epoch: str = None
+    version: str = None
+    release: str = None
+    repo: str = None
+    _state: _PackageState = _PackageState.ABSENT
+
+    @property
+    def installed(self) -> bool:
+        """Determine if package is marked 'installed'."""
+        return self._state == _PackageState.INSTALLED
+
+    @property
+    def available(self) -> bool:
+        """Determine if package is marked 'available'."""
+        return self._state == _PackageState.AVAILABLE
+
+    @property
+    def absent(self) -> bool:
+        """Determine if package is marked 'absent'."""
+        return self._state == _PackageState.ABSENT
+
+    @property
+    def full_version(self) -> Optional[str]:
+        """Get full version of package."""
+        if self.absent:
+            return None
+
+        full_version = [self.version, f"-{self.release}"]
+        if self.epoch:
+            full_version.insert(0, f"{self.epoch}:")
+
+        return "".join(full_version)
+
+
+def version() -> str:
+    """Get version of `dnf` executable."""
+    return _dnf("--version").splitlines()[0]
+
+
+def installed() -> bool:
+    """Determine if the `dnf` executable is available on PATH."""
+    return shutil.which("dnf") is not None
+
+
+def upgrade(*packages: Optional[str]) -> None:
+    """Upgrade one or more packages.
+
+    Args:
+        *packages (Optional[str]):
+            Packages to upgrade on system. If packages is omitted,
+            upgrade all packages on the system.
+    """
+    _dnf("upgrade", *packages)
+
+
+def install(*packages: Union[str, os.PathLike]) -> None:
+    """Install one or more packages.
+
+    Args:
+        *packages (Union[str, os.PathLike]): Packages to install on the system.
```

Comment:
Optional nit: the Google styleguide suggests you can omit data types in your args/returns docstrings where type annotations are used - personally I think it makes things easier/less messy to read.

```suggestion
        *packages: Packages to install on the system.
```

Reply:
This applies throughout :)

Reply:
Sounds good. I removed the type annotations from the docstrings in https://github.com/canonical/operator-libs-linux/pull/75/commits/9f137e07fe136b2ab085de263ad7d5c647b8c05d. I just really like knowing what my expected types are :nerd_face: